### PR TITLE
No var keyword rule

### LINKED ISSUE #378: new rule to disallow `var` declarations in favor of `let` and `const`

### PR DESCRIPTION
**Pull request by DickvdBrink**
Added no-var-keyword rule.

I couldn't think of a better name for this rule so if you have a better name I will change it :)
fix #378
Feel free to close this PR if this isn't the way you want it :)

Btw: I'm currently on vacation so feedback might take till around Sunday :)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -172,6 +172,7 @@ A sample configuration file with all options is available [here](https://github.
     * `"check-parameters"` disallows unused function and constructor parameters.
         * NOTE: this option is experimental and does not work with classes that use abstract method declarations, among other things. Use at your own risk.
 * `no-use-before-declare` disallows usage of variables before their declaration.
+* `no-var-keyword` disallows ussage of the `var` keyword, use `let` or `const` instead.
```
Comment (ash211) [inline README.md]: Typo: ussage -> usage

File: src/rules/noVarKeywordRule.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+        super.visitVariableStatement(node);
+        this.skipVariableDeclaration = false;
+    }
+
+    public visitVariableDeclaration(node: ts.VariableDeclaration) {
+
+        if (!this.skipVariableDeclaration) {
+            var flags = node.parent.flags;
+            var name = (<ts.Identifier> node.name);
```
Comment (adidahiya) [inline src/rules/noVarKeywordRule.ts]: `node.name` might be a `BindingPattern` (a.k.a. destructuring), so you'll have to handle both cases.

File: src/rules/noVarKeywordRule.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            this.hasModifier(node.modifiers, ts.SyntaxKind.DeclareKeyword)) {
+            this.skipVariableDeclaration = true;
+        }
+
+        super.visitVariableStatement(node);
+        this.skipVariableDeclaration = false;
+    }
+
+    public visitVariableDeclaration(node: ts.VariableDeclaration) {
+
```
Comment (adidahiya) [inline src/rules/noVarKeywordRule.ts]: remove this newline

File: src/rules/noVarKeywordRule.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        if (modifiers == null) {
+            return false;
+        }
+        for (var i = 0; i < modifiers.length; i++) {
+            var modifier = modifiers[i];
+            if (modifier.kind === modifierKind) {
+                return true;
+            }
+        }
+        return false;
```
Comment (adidahiya) [inline src/rules/noVarKeywordRule.ts]: L62–L68 can be similified:

```ts
return modifiers.some((modifier) => modifier.kind === modifierKind);
```

File: src/rules/noVarKeywordRule.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        return this.applyWithWalker(noVarKeywordWalker);
+    }
+}
+
+class NoVarKeywordWalker extends Lint.RuleWalker {
+    private skipVariableDeclaration = false;
+
+    public visitVariableStatement(node: ts.VariableStatement): void {
+        if (this.hasModifier(node.modifiers, ts.SyntaxKind.ExportKeyword) ||
+            this.hasModifier(node.modifiers, ts.SyntaxKind.DeclareKeyword)) {
```
Comment (adidahiya) [inline src/rules/noVarKeywordRule.ts]: please move the `||` operator to the next line

File: test/files/rules/novarkeyword.test.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+let bar;
+const qux;
+
+let k,
+    l;
+
+module quz {
+    export var i;
+}
```
Comment (adidahiya) [inline test/files/rules/novarkeyword.test.ts]: let's add some more tests for 1.5 syntax, like `export default`, `import`, destructuring (`var [a, b] = [1, 2]`)...

File: src/rules/noVarKeywordRule.ts (hunk 4 above)
Comment (adidahiya) [inline src/rules/noVarKeywordRule.ts]: After you rebase on the latest `master`, you can use the new util function `Lint.hasModifier` instead.

**Comment by adidahiya** [comment]
Any estimate on when you'll be able to make these fixes? Should I take up this PR?

**Comment by DickvdBrink** [comment]
Oops sorry, will fix some stuff tonight. Actually forgot about this a bit...

I do have a small question though!
The location where this error is thrown might be a bit odd, it is now thrown twice on the code below and on the location of the name but there is only one `var` keyword so that might be a bit weird. Do you agree?

    var i,
        j;

**Comment by adidahiya** [comment]
I agree, it's excessive to report 2 failures in the case of a variable declaration list of length 2 like that. You could try making the walker visit `VariableDeclarationList` nodes to get the desired behavior.

**Comment by DickvdBrink** [comment]
Fixed most of the feedback stuff (hopefully in the right way) and rebased on master.

What I didn't fix (yet) is the `export default` stuff because I don't know how to make it usable in a single TypeScript file..

**Pull request merged**